physics-integration-layer/src/math.rs

```diff
@@ -0,0 +1,154 @@
+#[cfg(not(feature = "f64"))]
+pub type TargetVec3 = Vec3;
+
+#[cfg(feature = "f64")]
+pub type TargetVec2 = DVec2;
+#[cfg(not(feature = "f64"))]
+pub type TargetVec2 = Vec2;
+
+#[cfg(feature = "f64")]
+pub type TargetQuat = DQuat;
+#[cfg(not(feature = "f64"))]
+pub type TargetQuat = Quat;
+
+// Taken from `bevy_xpbd` https://github.com/Jondolf/bevy_xpbd/blob/main/src/math/double.rs#L39
+/// Adjust the precision of the math construct to the precision chosen for compilation.
+pub trait AdjustPrecision {
+    /// A math construct type with the desired precision.
+    type Adjusted;
+    /// Adjusts the precision of [`self`] to [`Self::Adjusted`](#associatedtype.Adjusted).
+    fn adjust_precision(&self) -> Self::Adjusted;
+}
+
+impl AdjustPrecision for f32 {
+    type Adjusted = TargetFloat;
+    fn adjust_precision(&self) -> Self::Adjusted {
+        #[cfg(feature = "f64")]
+        return (*self).into();
+        #[cfg(not(feature = "f64"))]
+        return *self;
+    }
+}
+
+#[cfg(feature = "f64")]
+impl AdjustPrecision for f64 {
+    type Adjusted = TargetFloat;
+    fn adjust_precision(&self) -> Self::Adjusted {
+        #[cfg(feature = "f64")]
+        return *self;
+        #[cfg(not(feature = "f64"))]
+        return (*self).into();
+    }
+}
+
+impl AdjustPrecision for Vec3 {
+    type Adjusted = TargetVec3;
+    fn adjust_precision(&self) -> Self::Adjusted {
+        #[cfg(feature = "f64")]
+        return self.as_dvec3();
+        #[cfg(not(feature = "f64"))]
+        return *self;
+    }
+}
+
+#[cfg(feature = "f64")]
+impl AdjustPrecision for DVec3 {
+    type Adjusted = TargetVec3;
+    fn adjust_precision(&self) -> Self::Adjusted {
+        #[cfg(feature = "f64")]
+        return *self;
+        #[cfg(not(feature = "f64"))]
+        return self.as_vec3();
+    }
+}
+
+impl AdjustPrecision for Quat {
+    type Adjusted = TargetQuat;
+    fn adjust_precision(&self) -> Self::Adjusted {
+        #[cfg(feature = "f64")]
+        return self.as_dquat();
+        #[cfg(not(feature = "f64"))]
+        return *self;
+    }
+}
+
+#[cfg(feature = "f64")]
+impl AdjustPrecision for DQuat {
+    type Adjusted = TargetQuat;
+    fn adjust_precision(&self) -> Self::Adjusted {
+        #[cfg(feature = "f64")]
+        return *self;
+        #[cfg(not(feature = "f64"))]
+        return self.as_quat();
+    }
+}
+
+impl AdjustPrecision for Vec2 {
+    type Adjusted = TargetVec2;
+    fn adjust_precision(&self) -> Self::Adjusted {
+        #[cfg(feature = "f64")]
+        return self.as_dvec2();
+        #[cfg(not(feature = "f64"))]
+        return *self;
+    }
+}
+
+#[cfg(feature = "f64")]
+impl AdjustPrecision for DVec2 {
+    type Adjusted = TargetVec2;
+    fn adjust_precision(&self) -> Self::Adjusted {
+        #[cfg(feature = "f64")]
+        return *self;
+        #[cfg(not(feature = "f64"))]
+        return self.as_vec2();
+    }
+}
+
+/// Adjust the precision down to `f32` regardless of compilation.
+pub trait AsF32 {
```

Comment:
Why not merge this functionality into `AdjustPrecision`? Give it a second method - maybe `bevy_precision` or `lower_precision` - which brings it to the same precision Bevy uses.

Reply:
I was trying to follow the same way XPBD was using in their math library [here](https://github.com/Jondolf/bevy_xpbd/blob/e9b98d05be9e14086a8b6182d1bb55376cc978e4/src/math/mod.rs#L40), should I still integrate in the same Trait?

Reply:
In that case, let's just leave it as that for now.